binding/go/porcupine.go

```diff
@@ -0,0 +1,281 @@
+	FrameLength = nativeFrameLength()
+
+	// Audio sample rate accepted by Picovoice.
+	SampleRate = nativeSampleRate()
+
+	// Porcupine version
+	Version = nativeVersion()
+)
+
+// Init function for Porcupine. Must be called before attempting process
+func (porcupine *Porcupine) Init() (err error) {
+	if porcupine.ModelPath == "" {
+		porcupine.ModelPath = defaultModelFile
+	}
+
+	if _, err := os.Stat(porcupine.ModelPath); os.IsNotExist(err) {
+		return fmt.Errorf("%s: Specified model file could not be found at %s", pvStatusToString(INVALID_ARGUMENT), porcupine.ModelPath)
+	}
+
+	if porcupine.BuiltInKeywords != nil && len(porcupine.BuiltInKeywords) > 0 {
+		for _, keyword := range porcupine.BuiltInKeywords {
+			if err := keyword.IsValid(); err != nil {
+				return fmt.Errorf("%s: '%s' is not a valid built-in keyword.", pvStatusToString(INVALID_ARGUMENT), keyword)
+			}
+			keywordStr := string(keyword)
+			porcupine.KeywordPaths = append(porcupine.KeywordPaths, builtinKeywords[keywordStr])
+		}
+	}
+
+	if porcupine.KeywordPaths == nil || len(porcupine.KeywordPaths) == 0 {
+		return fmt.Errorf("%s: No valid keywords were provided.", pvStatusToString(INVALID_ARGUMENT))
+	}
+
+	if porcupine.Sensitivities == nil {
+		porcupine.Sensitivities = make([]float32, len(porcupine.KeywordPaths))
+		for i := range porcupine.KeywordPaths {
+			porcupine.Sensitivities[i] = 0.5
+		}
+	}
+
+	if len(porcupine.KeywordPaths) != len(porcupine.Sensitivities) {
+		return fmt.Errorf("%s: Keyword array size (%d) is not the same size as sensitivities array (%d)",
+			pvStatusToString(INVALID_ARGUMENT), len(porcupine.KeywordPaths), len(porcupine.Sensitivities))
+	}
+
+	ret := porcupine.nativeInit()
+	if PvStatus(ret) != SUCCESS {
+		return fmt.Errorf(": Porcupine returned error %s", pvStatusToString(INVALID_ARGUMENT))
+	}
+
+	return nil
+}
+
+// Releases resources acquired by Porcupine.
+func (porcupine *Porcupine) Delete() error {
+	if porcupine.handle == 0 {
+		return fmt.Errorf("Porcupine has not been initialized or has already been deleted.")
+	}
+
+	porcupine.nativeDelete()
+	return nil
+}
+
+// Processes a frame of the incoming audio stream and emits the detection result.
+// Frame of audio The number of samples per frame can be attained by calling
+// `.FrameLength`. The incoming audio needs to have a sample rate equal to `.Sample` and be 16-bit
+// linearly-encoded. Porcupine operates on single-channel audio.
+// Returns a 0 based index if keyword was detected in frame. Returns -1 if no detection was made.
+func (porcupine *Porcupine) Process(pcm []int16) (int, error) {
+
+	if porcupine.handle == 0 {
+		return -1, fmt.Errorf("Porcupine has not been initialized or has been deleted.")
+	}
+
+	if len(pcm) != FrameLength {
+		return -1, fmt.Errorf("Input data frame is wrong size")
+	}
+
+	// call process
+	ret, index := porcupine.nativeProcess(pcm)
+	if PvStatus(ret) != SUCCESS {
+		return -1, fmt.Errorf("Process audio frame failed with PvStatus: %d", ret)
```

Comment:
return the status instead of -1?

Reply:
first value is keyword index - I wouldn't want folks getting confused that they got a valid detection even if an error is returned. That's why I put the status code in the error.

Reply:
makes sense